torch/csrc/jit/codegen/cuda/runtime/welford.cu

```diff
@@ -0,0 +1,423 @@
+    __syncthreads();
+    if (tid < rblock_size) {
+      shared_buf_M2[tid] = inp_M2;
+      shared_buf_avg[tid] = inp_avg;
+    }
+    __syncthreads();
+    if (should_write) {
+      size_t offset_write =
+          offset_in_reduction_block<X_THREAD, Y_THREAD, Z_THREAD>(
+              threadIdx, blockDim);
+      inp_M2 = shared_buf_M2[offset_write];
+      inp_avg = shared_buf_avg[offset_write];
+    }
+  }
+
+  if (should_write && read_write_pred) {
+    out_M2 = inp_M2;
+    out_avg = inp_avg;
+    out_N = inp_N;
+  }
+}
+
+/**
+    Grid welford combine (WIP)
+*/
+template <
+    bool X_BLOCK,
+    bool Y_BLOCK,
+    bool Z_BLOCK,
+    bool X_THREAD,
+    bool Y_THREAD,
+    bool Z_THREAD,
+    typename T>
+__device__ bool gridWelford(
+    T& out_M2,
+    T& out_avg,
+    int& out_N,
+    T inp_M2,
+    T inp_avg,
+    int inp_N,
+    volatile T* work_buf_M2,
+    volatile T* work_buf_avg,
+    Tensor<int64_t, 1> sync_flags,
+    T* shared_buf_M2,
+    T* shared_buf_avg,
+    bool read_write_pred,
+    T init_val) {
+  // Number of values to reduce in the grid dimensions
+  const auto seg_size =
+      size_of_reduction_segment<X_BLOCK, Y_BLOCK, Z_BLOCK>(gridDim);
+
+  // Index of the reduction we're performing out of the seg_size
+  const auto seg_idx =
+      index_of_reduction_segment<X_BLOCK, Y_BLOCK, Z_BLOCK>(blockIdx, gridDim);
+
+  // Number of threads we can use in final reduction, Seems to assume all
+  // threads in the block participate
+  const auto rblock_size =
+      size_of_reduction_block<X_THREAD, Y_THREAD, Z_THREAD>(blockDim);
+
+  // advance to the offset for this segment
+  // index of reduction * size of the reduction * size of threads
+  shared_buf_M2 += seg_idx * seg_size * rblock_size;
+  shared_buf_avg += seg_idx * seg_size * rblock_size;
+  if ((X_THREAD || threadIdx.x == 0) && (Y_THREAD || threadIdx.y == 0) &&
+      (Z_THREAD || threadIdx.z == 0)) {
+    auto rblock_offset = offset_in_reduction_segment<X_BLOCK, Y_BLOCK, Z_BLOCK>(
+        blockIdx, gridDim);
+    auto thread_offset =
+        offset_in_reduction_block<X_THREAD, Y_THREAD, Z_THREAD>(
+            threadIdx, blockDim);
+    auto work_buf_offset = rblock_size * rblock_offset + thread_offset;
+    if (read_write_pred) {
+      work_buf_M2[work_buf_offset] = inp_M2;
+      work_buf_avg[work_buf_offset] = inp_avg;
+    } else {
+      work_buf_M2[work_buf_offset] = init_val;
+      work_buf_avg[work_buf_offset] = init_val;
+    }
+  }
+  __syncthreads();
+
+  __shared__ bool last_block;
+  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
+    __threadfence();
+    auto old = (int64_t)atomicAdd((unsigned long long*)&sync_flags[seg_idx], 1);
+    last_block = old + 1 == seg_size;
+  }
+  __syncthreads();
+
+  if (last_block) {
+    // final reduction
+    gridWelfordLastBlock<X_THREAD, Y_THREAD, Z_THREAD>(
+        out_M2,
+        out_avg,
+        out_N,
+        (T*)work_buf_M2,
+        (T*)work_buf_avg,
+        inp_N,
```

Comment:
similarly, can we assume consistent `inp_N` across grid?

Reply:
Great question. I am currently assume so since currently I'm only supporting welford on a single tensor instead of user supporting arbitrary array of var/avg/count. In the case if we construct the var/avg/count from a raw data tensor, the count values should be consistent.

Reply:
> raw data tensor instead of user supporting arbitrary tensors

What's the difference between the two? I think this question also applies to `N` in the argument of `blockWelford` -> How would `raw data tensor` ensure each thread/block cover the same pixels during both stage (serial / block reduction).

Reply:
Actually good catch. Thanks for pointing this out.  

On a second thought this is a potential problem that I didn't cover in the test. This optimization to use all the threads for the final reduction was borrowed from grid_reduction : https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu#L166-L171

But apparently it doesn't apply well here since extra threads actually introduces in-consistent N across threads. But I believe we do want to use all the threads for the final reduction so I will refactor this part.

Reply:
On the higher level interfaces I'd assume the N's consistent if the function will correspond to a Fusion IR op. The use model is similar to a reduction op, say 
```cpp
T1_var, T1_avg, T1_N = welford(T0[I0,I1,I2],0,1);
``` 
Then entries in T1_N would be `T0.size[0] * T0.size[1]` for all the output pixels.